### PR TITLE
Add the PolyCheckedVector type

### PR DESCRIPTION
**Pull request by cjpatton**
Partially addresses #23.

This change adds a new type called `PolyCheckedVector`. This type represents vectors for which `p(x) == 0` for some polynomial `p` and each vector element `x`. It can be used to construct higher level types, such as boolean vectors. The construction is based on [BBG+19, Theorem 5.3]. Its performance is comparable to the proof system of Prio v2, though proof verification is generally faster. 

Here are the numbers for proof generation, on my laptop (sizes are in number of field elements):

| proof system (input size) | average prove time | average query time | proof size |
|-------------------|------------|-----------|--------------|
| Prio v2 (10) | 7.1 us | 8.2 us | 29 |
| PolyCheckedVector (10) | 6.7 us | 5.9 us | 32 | 
| Prio v2 (100) | 47.3 us | 39.5 us | 231 |
| PolyCheckedVector (100) | 45.9 us | 30.7 us | 256 |
| Prio v2 (1,000) | 0.46 ms | 0.35 ms | 2,027 |
| PolyCheckedVector (1,000) | 0.45 ms | 0.26 ms | 2,048 |
| Prio v2 (10,000) | 10.4 ms | 7.8 ms | 26,387 |
| PolyCheckedVector (10,000) | 11.4 ms | 6.2 ms | 32,768 | 

You can run these benchmarks (all except input size of 1mil) with `cargo bench`. Note that I refactored the code in `client.rs` a bit in order to expose a suitable benchmark for v2.

[[UPDATE: Originally the `PolyCheckedVector` type employed the "parallel sum" technique of Corollary 4.9 in order to reduce the proof size to `O(sqrt(n))`. I now realize that this technique can't be used here directly.]]

### CODE REVIEW

**Comment by cjpatton** [comment]
It occurs to me that there might be a way for me to reduce the overhead of my code somewhat. I'm moving this PR back to WIP.

**Comment by cjpatton** [comment]
Managed to shave a bit off. That's enough.

File: src/pcp/gadgets.rs [hunk 1]
```diff
@@ -2,42 +2,69 @@
...
+use std::marker::PhantomData;
+
+const FFT_THRESHOLD_MUL: usize = 60;
+const FFT_THRESHOLD_POLY_EVAL: usize = 8;
 
 /// An arity-2 gadget that multiples its inputs.
 pub struct Mul<F: FieldElement> {
-    buf: Vec<F>,
+    n: usize, // Size of buffer for FFT operations
+    n_inv: F, // Inverse of n in F
```
Comment (tgeoghegan) [inline src/pcp/gadgets.rs:19]:
```suggestion
    /// Size of buffer for FFT operations
    n: usize,
    // Inverse of n in F
    n_inv: F,
```

File: src/pcp/gadgets.rs [hunk 2]
```diff
@@ -2,42 +2,69 @@
...
-use crate::fft::{discrete_fourier_transform, discrete_fourier_transform_inv};
+use crate::fft::{discrete_fourier_transform, discrete_fourier_transform_inv_finish};
 use crate::field::FieldElement;
 use crate::pcp::{Gadget, GadgetCallOnly, PcpError};
+use crate::polynomial::{poly_deg, poly_eval, poly_mul};
+
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+
+const FFT_THRESHOLD_MUL: usize = 60;
```
Comment (tgeoghegan) [inline src/pcp/gadgets.rs:13]: I saw your comment in `speed_tests.rs` explaining how these values were derived. I think we should have a similar comment here.

File: src/pcp/gadgets.rs [hunk 3]
```diff
@@ -49,25 +76,186 @@ impl<F: FieldElement> GadgetCallOnly<F> for Mul<F> {
...
 
+/// An arity-1 gadget that evaluates its input on some polynomial.
+pub struct PolyEval<F: FieldElement> {
+    poly: Vec<F>,
+    n: usize, // Size of buffer for FFT operations
+    n_inv: F, // Inverse of n in F
+}
+
+impl<F: FieldElement> PolyEval<F> {
+    /// Returns a gadget that evaluates its input on `poly`. The prameter `in_len` denotes the
```
Comment (tgeoghegan) [inline src/pcp/gadgets.rs:99]:
```suggestion
    /// Returns a gadget that evaluates its input on `poly`. The parameter `in_len` denotes the
```

File: src/pcp/gadgets.rs [hunk 4]
```diff
@@ -49,25 +76,186 @@ impl<F: FieldElement> GadgetCallOnly<F> for Mul<F> {
...
+/// outputs. The arity is equal to the arity of the inner gadget times the number of chunks.
+pub struct ParallelSum<F: FieldElement, G: Gadget<F>> {
+    inner: G,
+    chunks: usize,
+    phantom: PhantomData<F>,
+}
+
+impl<F: FieldElement, G: Gadget<F>> ParallelSum<F, G> {
+    /// Wraps `inner` into a parallel sum gadget with `chunks` chunks.
+    pub fn new(inner: G, chunks: usize, _in_len: usize) -> Self {
```
Comment (tgeoghegan) [inline src/pcp/gadgets.rs:201]: If `in_len` is unused, why not remove it from the prototype?

File: src/pcp/gadgets.rs [hunk 5]
```diff
@@ -49,25 +76,186 @@ impl<F: FieldElement> GadgetCallOnly<F> for Mul<F> {
...
+    }
+}
+
+impl<F: FieldElement, G: Gadget<F>> GadgetCallOnly<F> for ParallelSum<F, G> {
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        gadget_call_check(self, inp.len())?;
+        let mut outp = F::zero();
+        for chunk in inp.chunks(self.inner.call_in_len()) {
+            outp += self.inner.call(chunk)?;
+        }
```
Comment (tgeoghegan) [inline src/pcp/gadgets.rs:216]: It seems like the computation on each chunk is totally independent. Would it be worthwhile to actually make this parallel? Perhaps we should write up an issue for that future performance optimization.

File: src/pcp/types.rs [hunk 6]
```diff
@@ -74,6 +74,107 @@ impl<F: FieldElement> Value<F, Mul<F>> for Boolean<F> {
...
+    fn valid(&self, g: &mut dyn GadgetCallOnly<F>, rand: &[F]) -> Result<F, PcpError> {
+        if rand.len() < 1 {
+            return Err(PcpError::CircuitIn("random input not long enough"));
+        }
+
+        let mut r = rand[0];
+        let mut outp = F::zero();
+        for chunk in self.data.chunks(self.chunk_len) {
+            let n = chunk.len();
+            let v = match n == self.chunk_len {
```
Comment (tgeoghegan) [inline src/pcp/types.rs:124]: This works fine if you want to keep it but you could make this a little less indented if you do 
```
let v = if n == self.chunk_len {
    g.call(chunk)?
} else {
    // the other thing
};
```

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: src/pcp/gadgets.rs (hunk 4 above)
Comment (cjpatton) [reply src/pcp/gadgets.rs:201]: Ugh, good catch. That's just left over from an earlier rewrite.

File: src/pcp/gadgets.rs (hunk 5 above)
Comment (cjpatton) [reply src/pcp/gadgets.rs:216]: Done: https://github.com/abetterinternet/libprio-rs/issues/34

File: src/pcp/types.rs (hunk 6 above)
Comment (cjpatton) [reply src/pcp/types.rs:124]: Done.

**Review by tgeoghegan — APPROVED** [review]

**Pull request merged**